operators/src/processing/raster_kernel/kernel_function.rs

```diff
@@ -0,0 +1,208 @@
+use super::{error, RasterKernelError};
+use crate::util::{number_statistics::NumberStatistics, Result};
+use geoengine_datatypes::raster::{Grid2D, GridShape2D, GridSize, Pixel};
+use num::Integer;
+use num_traits::AsPrimitive;
+use snafu::ensure;
+
+/// A function that calculates one pixel values from a neighborhood of pixels.
+pub trait KernelFunction<P>: Sync + Send + Clone
+where
+    P: Pixel,
+    f64: AsPrimitive<P>,
+{
+    fn kernel_size(&self) -> GridShape2D;
+
+    /// Specifies the x extent beginning from the center pixel
+    fn x_radius(&self) -> usize {
+        self.kernel_size().axis_size_x() / 2
+    }
+
+    fn x_width(&self) -> usize {
+        self.kernel_size().axis_size_x()
+    }
+
+    /// Specifies the y extent beginning from the center pixel
+    fn y_width(&self) -> usize {
+        self.kernel_size().axis_size_y()
+    }
+
+    /// Specifies the x extent right of one pixel
+    fn y_radius(&self) -> usize {
+        self.kernel_size().axis_size_y() / 2
+    }
+
+    fn apply(&self, values: &[Option<P>]) -> Option<P>;
+}
+
+/// A kernel that computes the standard deviation of a neighborhood of pixels.
+#[derive(Debug, Clone, Copy)]
+pub struct StandardDeviationKernel {
+    dimensions: GridShape2D,
+}
+
+impl StandardDeviationKernel {
+    pub fn new(dimensions: GridShape2D) -> Result<Self, RasterKernelError> {
+        ensure!(
+            dimensions.axis_size_x().is_odd() && dimensions.axis_size_y().is_odd(),
+            error::DimensionsNotOdd { actual: dimensions }
+        );
+
+        Ok(Self { dimensions })
+    }
+}
+
+impl<P> KernelFunction<P> for StandardDeviationKernel
+where
+    P: Pixel,
+    f64: AsPrimitive<P>,
+{
+    fn kernel_size(&self) -> GridShape2D {
+        self.dimensions
+    }
+
+    fn apply(&self, values: &[Option<P>]) -> Option<P> {
+        if values.len() != self.dimensions.number_of_elements() {
+            return None;
+        }
+
```

Comment:
isn't the NumerStatistics built with huge number of elements in mind? Here, we know we will only ever have a handful of elements. Would it make sense to use something else here? If so, maybe add a TODO

Reply:
The kernel can be quite large, so I think this is okay for now.
If the size was fixed, I guess this would make sense to optimize now.